internal/state/enumeratedtotal/tree.go

```diff
@@ -206,3 +210,69 @@ func headRoot(contract common.Address, store storage.Tx) (common.Hash, error) {
 func setHeadRoot(contract common.Address, store storage.Tx, root common.Hash) error {
 	return store.Set([]byte(headRootKeyPrefix+contract.String()), root.Bytes())
 }
+
+// TagRoot stores a root value for the block so that it can be checked later
+func (b *tree) TagRoot(blockNumber int64) error {
+	tagKey := tagPrefix + b.contract.String() + "/" + strconv.FormatInt(blockNumber, 10)
+	root := b.Root()
+	err := b.store.Set([]byte(tagKey), root.Bytes())
+	if err != nil {
+		return err
+	}
+
+	totalSupply, err := b.TotalSupply()
+	if err != nil {
+		return err
+	}
+	tagTotalSupplyKey := totalSupplyTagPrefix + b.contract.String() + "/" + strconv.FormatInt(blockNumber, 10)
+	return b.store.Set([]byte(tagTotalSupplyKey), []byte(strconv.FormatInt(totalSupply, 10)))
+}
+
+// Checkout sets the current root to the one that is tagged for a blockNumber.
```

Comment:
maybe we could add a TODO to extract some functionality that is repeated (in the other trees) to a helper function

Reply:
yes, I was thinking that maybe that makse sense. There is some code that is repetitive but not all. I will add TODO